### PR TITLE
CI cleanup; All CI workflows are required to pass before merge to develop.

### LINKED ISSUE #242: improve CI system
- Create a developer build which does FTP testing, address santizer, doxygen build, warning check, code coverage, and Debug build. Turn these settings off in all other builds. Use develop branch of all required NCEPLIBS libraries and the latest version of jasper.
- Create a linux_versions workflow that checks with different versions of dependencies.
- Create a linux_settings workflow that tests different build options with the most recent versions of dependencies.
- Use ubuntu-latest/macos-latest everywhere.
- Use caches everywhere they can be used, including on code checkouts.

### PR DESCRIPTION
**Pull request by edwardhartnett**
Fixes #242 

Cleanup of CI system. All CI workflows are required to pass before merge to develop.

### CODE REVIEW

**Comment by edwardhartnett** [comment]
Code coverage is working on developer workflow.

**Comment by edwardhartnett** [comment]
The new Intel build is failing like this:

```
6: ImportError: /home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/build/install/lib/python3.1/site-packages/_bufrlib.cpython-310-x86_64-linux-gnu.so: undefined symbol: __intel_sse2_strcpy
 6/62 Test  #6: test_pyncepbufr_write ............***Failed    0.09 sec

```
Anyone have any idea what is going on?

**Comment by jbathegit** [comment]
This has been an issue for a long time - see #231.  That in turn references #135, where I mentioned this problem back in August of 2021.  The python tests have never worked correctly under Intel, regardless of whether or not the code coverage flags are on.

**Comment by edwardhartnett** [comment]
OK, I will turn off Python testing with Intel...

**Comment by edwardhartnett** [comment]
@jbathegit if this looks good to you, approve it and get it merged. I had to change the settings to match the names used in this PR, so nothing else can be merged without updating their CI files to these. Let me know if that's a problem...

**Comment by jbathegit** [comment]
Thanks for adding in the Intel test, as it's long been a goal of mine to include that as one of the CI tests, but I could never figure out any way to do that!

A question though - is there any particular reason why the python builds and tests are now turned off for the MacOS?  Under the old main.yml, we did have python builds and tests included on MacOS, as well as on all of the various Linux flavors.

**Comment by edwardhartnett** [comment]
@jbathegit Good catch, I will turn on the python tests with MacOS...

File: .github/workflows/MacOS.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      with: 
+        path: bufr
+
+    - name: build-bufr
+      run: |
+        cd bufr
+        mkdir build 
+        cd build
+        cmake ..
+        make -j2 VERBOSE=1
```
Comment (jbathegit) [inline .github/workflows/MacOS.yml]: lines 41 and 42 are now redundant and should be removed

File: .github/workflows/MacOS.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        cmake -DCMAKE_INSTALL_PREFIX=./install -DENABLE_PYTHON=ON ..
+        make -j2 VERBOSE=1
+        make install
+        cd python
+        sudo python3 setup.py install
+
+    - name: test-bufr
+      run: |
+        cd bufr/build
```
Comment (jbathegit) [inline .github/workflows/MacOS.yml:49]: I noticed in some of the yml files this was just `cd bufr/build` and in others it was `cd $GITHUB_WORKSPACE/bufr/build`.  Which one is correct?

Even if they both work, I think consistency among all of them would be helpful ;-)

File: .github/workflows/MacOS.yml (hunk 1 above)
Comment (edwardhartnett) [reply .github/workflows/MacOS.yml]: Good catch.

File: .github/workflows/MacOS.yml (hunk 2 above)
Comment (edwardhartnett) [reply .github/workflows/MacOS.yml:49]: Fixed.

**Comment by jbathegit** [comment]
I see where in commit 32d0131 you noted you were "taking python out of develop as it clashes with asan".  What is "asan"?

**Comment by edwardhartnett** [comment]
asan is short for the address santizer. It's turned on in develop with -fsanitize=address in C and Fortran flags.

The asan run is great because it catches memory errors and tells you the exact line of the code where the problem is. Super-helpful! But it does not play well with other tools. For example, using the debugger with code that has been built with address santizer causes a similar error to that seen here, a complaint about asan. 

The solution is to do one asan build in develop. I would love to also do a python build in develop, but we are already doing that in Linux, so we can take it out of develop.

**Comment by jbathegit** [comment]
Ahh, OK thanks for clarifying that.  I did see where you'd opened up #244, and that's a bit of a head-scratcher.  Jack or I will have to take a good look at that at some point, but for now I want to get some other stuff tidied up, including #78 which we're very close to resolving.  In the meantime, I'll go ahead and get this merged in.

**Review by jbathegit — APPROVED** [review]

**Pull request merged**